Prod deploy failed on main

## Prod deploy failed

**Repo**: alexsiri7/reli
**SHA**: `56b7a8270982c940ff38d99bf69e0c26cb2b2ead`
**Deploy run/status**: https://github.com/alexsiri7/reli/actions/runs/25004397069
**State**: failure
**Deployed at**: 2026-04-27T15:34:41Z

Auto-filed by `pipeline-health-cron.sh`. Prod deploy is a pipeline bottleneck — archon has been fired immediately. This issue is tagged `archon:in-progress` so the pickup cron will not double-fire.

### Steps
1. Inspect deploy logs at https://github.com/alexsiri7/reli/actions/runs/25004397069
2. Identify root cause (Railway config? env var? build error? migration?)
3. Fix on a branch, land, confirm next deploy goes green

## 🔍 Investigation: Prod deploy failed on main

**Type**: `BUG`

### Assessment

| Metric | Value | Reasoning |
|--------|-------|-----------|
| Severity | `HIGH` | Prod pipeline blocked — every push to main fails to deploy until token is valid |
| Complexity | `LOW` | One workflow step to extend in two places; no backend/frontend code changes |
| Confidence | `HIGH` | Run log directly shows `"Not Authorized"` on `serviceInstanceUpdate`; timeline confirms token expired at 15:07 UTC, deploy ran at 15:34 UTC |

---

### Problem Statement

The `RAILWAY_TOKEN` GitHub secret holds an expired (or otherwise invalid) Railway API token. The "Validate Railway secrets" step (added in PR #735) only checks that the secret is **non-empty** — it does not verify the token is accepted by Railway. So the pipeline passes validation, then fails 4 seconds later with a cryptic `"Not Authorized"` error in the deploy step.

---

### Root Cause Analysis

WHY the deploy step fails:
↓ BECAUSE Railway GraphQL `serviceInstanceUpdate` returns `{"errors":[{"message":"Not Authorized"...}]}`

↓ BECAUSE the `RAILWAY_TOKEN` passed as `Authorization: Bearer` is rejected by Railway

↓ BECAUSE the token expired at 15:07 UTC 2026-04-27, and the deploy ran at 15:34 UTC
  — commit `56b7a82` says "token was rotated and CI has been green since", but the deploy triggered by that very commit failed at 15:34:41Z

↓ **ROOT CAUSE**: `RAILWAY_TOKEN` secret contains an expired/invalid token. The pre-flight validation only checks `[ -z "$RAILWAY_TOKEN" ]` (non-empty), not that Railway actually accepts it.

Evidence from run log:
```
serviceInstanceUpdate response: {"errors":[{"message":"Not Authorized",...}],"data":null}
jq: error (at <stdin>:1): Railway API error: Not Authorized
##[error]Process completed with exit code 5.
```

---

### Implementation Plan

**Step 1 (Operational)**: Rotate `RAILWAY_TOKEN`

```bash
# Create a permanent token at https://railway.com/account/tokens
gh secret set RAILWAY_TOKEN --repo alexsiri7/reli
```

Verify it works before triggering the pipeline:
```bash
curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
  -H "Authorization: Bearer <new-token>" \
  -H "Content-Type: application/json" \
  -d '{"query":"{me{id}}"}' | jq .
# Expected: {"data":{"me":{"id":"..."}}}
```

**Step 2 (Code)**: Add token auth pre-flight check to both validate steps in `.github/workflows/staging-pipeline.yml`

| Step | File | Lines | Change |
|------|------|-------|--------|
| 1 | `.github/workflows/staging-pipeline.yml` | 32-48 | Add `me { id }` API check after missing-secrets guard |
| 2 | `.github/workflows/staging-pipeline.yml` | 139-155 | Same addition for production validate step |

<details>
<summary>📋 Detailed implementation — what to add to each validate step</summary>

Add this block **after** the existing `if [ -n "$missing" ]; then … fi` in both validate steps:

```bash
RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
  -H "Authorization: Bearer $RAILWAY_TOKEN" \
  -H "Content-Type: application/json" \
  -d '{"query":"{me{id}}"}')
if echo "$RESP" | jq -e '.errors' > /dev/null 2>&1; then
  echo "::error::RAILWAY_TOKEN is invalid or expired: $(echo "$RESP" | jq -r '.errors[0].message')"
  echo "Rotate the token — see DEPLOYMENT_SECRETS.md Token Rotation section."
  exit 1
fi
```

This gives a clear, actionable error ("RAILWAY_TOKEN is invalid or expired") instead of the cryptic "Not Authorized" buried in deploy step output.

</details>

---

### Validation

```bash
gh workflow run staging-pipeline.yml --repo alexsiri7/reli
```

Confirm green pipeline after token rotation + code change.

---

### Next Step

To implement: `/implement-issue 737`

---
*Investigated by Claude • 2026-04-27T15:55:00Z*

--- a/.github/workflows/staging-pipeline.yml
+++ b/.github/workflows/staging-pipeline.yml
@@ -46,6 +46,15 @@ jobs:
             echo "See DEPLOYMENT_SECRETS.md for setup instructions."
             exit 1
           fi
+          RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
+            -H "Authorization: Bearer $RAILWAY_TOKEN" \
+            -H "Content-Type: application/json" \
+            -d '{"query":"{me{id}}"}')
+          if echo "$RESP" | jq -e '.errors' > /dev/null 2>&1; then
+            echo "::error::RAILWAY_TOKEN is invalid or expired: $(echo "$RESP" | jq -r '.errors[0].message')"
+            echo "Rotate the token — see DEPLOYMENT_SECRETS.md Token Rotation section."
+            exit 1
+          fi
 
       - name: Deploy staging image to Railway
         env:
@@ -151,6 +160,15 @@ jobs:
           if [ -n "$missing" ]; then
             echo "::error::Missing required secrets:$missing"
             echo "See DEPLOYMENT_SECRETS.md for setup instructions."
+            exit 1
+          fi
+          RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
+            -H "Authorization: Bearer $RAILWAY_TOKEN" \
+            -H "Content-Type: application/json" \
+            -d '{"query":"{me{id}}"}')
+          if echo "$RESP" | jq -e '.errors' > /dev/null 2>&1; then
+            echo "::error::RAILWAY_TOKEN is invalid or expired: $(echo "$RESP" | jq -r '.errors[0].message')"
+            echo "Rotate the token — see DEPLOYMENT_SECRETS.md Token Rotation section."
             exit 1
           fi
 

--- a/.github/workflows/staging-pipeline.yml
+++ b/.github/workflows/staging-pipeline.yml
@@ -46,12 +46,13 @@ jobs:
             echo "See DEPLOYMENT_SECRETS.md for setup instructions."
             exit 1
           fi
-          RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
+          RESP=$(curl -s -X POST "https://backboard.railway.app/graphql/v2" \
             -H "Authorization: Bearer $RAILWAY_TOKEN" \
             -H "Content-Type: application/json" \
             -d '{"query":"{me{id}}"}')
-          if echo "$RESP" | jq -e '.errors' > /dev/null 2>&1; then
-            echo "::error::RAILWAY_TOKEN is invalid or expired: $(echo "$RESP" | jq -r '.errors[0].message')"
+          if ! echo "$RESP" | jq -e '.data.me.id' > /dev/null 2>&1; then
+            MSG=$(echo "$RESP" | jq -r '.errors[0].message // "could not reach Railway API or token rejected"')
+            echo "::error::RAILWAY_TOKEN is invalid or expired: $MSG"
             echo "Rotate the token — see DEPLOYMENT_SECRETS.md Token Rotation section."
             exit 1
           fi
@@ -162,12 +163,13 @@ jobs:
             echo "See DEPLOYMENT_SECRETS.md for setup instructions."
             exit 1
           fi
-          RESP=$(curl -sf -X POST "https://backboard.railway.app/graphql/v2" \
+          RESP=$(curl -s -X POST "https://backboard.railway.app/graphql/v2" \
             -H "Authorization: Bearer $RAILWAY_TOKEN" \
             -H "Content-Type: application/json" \
             -d '{"query":"{me{id}}"}')
-          if echo "$RESP" | jq -e '.errors' > /dev/null 2>&1; then
-            echo "::error::RAILWAY_TOKEN is invalid or expired: $(echo "$RESP" | jq -r '.errors[0].message')"
+          if ! echo "$RESP" | jq -e '.data.me.id' > /dev/null 2>&1; then
+            MSG=$(echo "$RESP" | jq -r '.errors[0].message // "could not reach Railway API or token rejected"')
+            echo "::error::RAILWAY_TOKEN is invalid or expired: $MSG"
             echo "Rotate the token — see DEPLOYMENT_SECRETS.md Token Rotation section."
             exit 1
           fi

